pageserver/src/import_datadir.rs

```diff
@@ -617,3 +622,94 @@ async fn read_all_bytes(reader: &mut (impl AsyncRead + Unpin)) -> Result<Bytes>
     reader.read_to_end(&mut buf).await?;
     Ok(Bytes::from(buf))
 }
+
+// Set this limit to a small value so that we are a
+// good async citizen and yield repeatedly (but not
+// too often for the small writes)
+const YIELD_DIST: u64 = 1024;
+
+/// A buffer that yields writes at repeated intervals
+struct YieldingVec {
+    yld_len: u64,
+    buf: Vec<u8>,
+}
+
+impl YieldingVec {
+    fn new() -> Self {
+        Self {
+            yld_len: 0,
+            buf: Vec::new(),
+        }
+    }
+    fn should_yield(&mut self, buf_len: usize) -> bool {
+        let buf_len: u64 = buf_len.try_into().unwrap();
+        let ret = self.yld_len / YIELD_DIST != (self.yld_len + buf_len) / YIELD_DIST;
+        self.yld_len += buf_len;
+        ret
+    }
+}
+
+impl AsyncWrite for YieldingVec {
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        _cx: &mut task::Context<'_>,
+        buf: &[u8],
+    ) -> Poll<std::io::Result<usize>> {
+        if self.should_yield(buf.len()) {
+            return Poll::Pending;
+        }
+        self.get_mut().buf.extend_from_slice(buf);
+        Poll::Ready(Ok(buf.len()))
+    }
+
+    fn poll_write_vectored(
+        mut self: Pin<&mut Self>,
+        _: &mut task::Context<'_>,
+        bufs: &[IoSlice<'_>],
+    ) -> Poll<std::io::Result<usize>> {
+        if self.should_yield(bufs.iter().map(|b| b.len()).sum()) {
+            return Poll::Pending;
+        }
+        Poll::Ready(std::io::Write::write_vectored(&mut self.buf, bufs))
+    }
+
+    fn is_write_vectored(&self) -> bool {
+        true
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, _cx: &mut task::Context<'_>) -> Poll<std::io::Result<()>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn poll_shutdown(
+        self: Pin<&mut Self>,
+        _cx: &mut task::Context<'_>,
+    ) -> Poll<std::io::Result<()>> {
+        Poll::Ready(Ok(()))
+    }
+}
+
+pub async fn create_tar_zst(pgdata_path: &Path) -> Result<Vec<u8>> {
+    let mut paths = Vec::new();
+    for entry in WalkDir::new(pgdata_path) {
+        let entry = entry?;
+        let metadata = entry.metadata().expect("error getting dir entry metadata");
+        if metadata.is_file() {
+            let path = entry.path().strip_prefix(pgdata_path)?;
+            paths.push(path.to_owned());
+        }
+    }
+    // Don't rely on file system order for listing as it may be non-deterministic
+    paths.sort();
+    let zstd = async_compression::tokio::write::ZstdEncoder::new(YieldingVec::new());
+    let mut builder = Builder::new(zstd);
+    // Use reproducible header mode
+    builder.mode(HeaderMode::Deterministic);
+    for path in paths {
+        builder.append_path(path).await?;
+    }
+    builder.finish().await?;
+    let zstd = builder.into_inner().await?;
+    let compressed = zstd.into_inner();
+    Ok(compressed.buf)
```

Comment:
You have an expectation how large this will be, could we add a logging if it is larger than your expectation?

Reply:
Ideally we get it down to one point something MB: https://github.com/neondatabase/neon/issues/2592#issuecomment-1720817221

Reply:
Done, see commit "Add warning for too large initdb.tar.zst"

Reply:
Common initdb.tar.zst sizes from running the tests with this commit:

```
1471339 nov 14 12:11 initdb.tar.zst
1471079 nov 14 12:11 initdb.tar.zst
```

The commit warns for archives larger than `2_000_000` bytes.